Port the compiler gym service backends to the new API

ChrisCummins opened the pull request:
Issue #254.

Commit:
[examples] Port the example services to use the new service runtime.

Issue #254.

Commit:
[llvm] Port LLVM service to the new CompilationSession interface.

Issue #254.

Commit:
[env] Add a fallback fork() implementation.

Supporting the Fork() operator is optional for CompilationSessions, so
provide a fallback implementation that creates a new environment and
replays the action sequence by hand.

Commit:
[rpc] Enable arena codegen.

hughleat commented on examples/example_compiler_gym_service/service_cc/ExampleService.cc:
```diff
@@ -2,23 +2,20 @@
+#include "compiler_gym/util/Unreachable.h"
 
-namespace compiler_gym::example_service {
+const char* usage = R"(Example CompilerGym service)";
```
Unnecessary use of R string?

ChrisCummins replied:
👍

hughleat commented on examples/example_compiler_gym_service/service_cc/ExampleService.cc:
```diff
@@ -29,179 +26,102 @@ template <typename T>
-}
+  std::vector<ActionSpace> getActionSpaces() const final override {
+    ActionSpace space;
+    space.set_name("default");
```
Shame these are protobufs, then they could just be params to the constructor.

ChrisCummins replied:
The problem is that this class is never instantiated by the user. They pass it into `CompilerGymService<T>` which handles lifetime management: 
https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/service/runtime/CompilerGymServiceImpl.h#L64

hughleat commented on examples/example_compiler_gym_service/env_tests.py:
```diff
@@ -27,6 +30,37 @@ def env(request) -> CompilerEnv:
+    scope="module",
+    params=[example.EXAMPLE_CC_SERVICE_BINARY, example.EXAMPLE_PY_SERVICE_BINARY],
+)
+def bin(request) -> Path:
```
Is this function unused?

ChrisCummins replied:
`@pytest.fixture` decorator declares a fixture, which means that any test function can then take `bin` as an argument and it will run and return the fixture value

hughleat reviewed (approved):
LGTM, although I still think that not making the user know anything about what RPC mechanism you are using would be better, including allowing a simpler API for them.

ChrisCummins commented:
Thanks for the review @hughleat!

The CI is complaining of a problem that I've never seen before. An undefined reference to the destructor of `google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()` and `google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()` in `getSpaces()`:

```
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
```

Will investigate and fix before merging.

Commit:
Bump bazel gRPC dependency to 1.37.1.

Commit:
Run `pre-commit install` as part of `make init`.

Commit:
[ci] Remove unnecessary env variables from build steps.

Commit:
[makefile] Add a BUILD_TARGET variable.

Commit:
[makefile] Make `make install-test` rules more flexible.

Encourage re-use of code within rules, and allow the test target to be
overriden from the command line.

Commit:
[tests] Fix fork() test.

process.poll() may return 0.

Commit:
[tests] Update llvm-stress fuzzer to new dataset API.

Commit:
[rpc] Raise ServiceError on service does not close gracefully.

Use SIGTERM rather than SIGKILL to terminate backend services, and
then check that the process ended with returncode zero.

Commit:
[service] Add graceful shutdown handling to gym services.

This adds a SIGTERM handler which enables the gRPC server to be
shutdown gracefully. This enables an address sanitized build to run
the end-of-process leak check.

Commit:
[ci] Add an asan test for LLVM service.

Commit:
[rpc] Make session leak detection consistent between C++/Python.

Commit:
[tests] Update the connection tests.

Commit:
[tests] Update LLVM env tests.

Commit:
[rpc] Raise a single exception on sessions leak check failure.

Commit:
[env] Log a failed attempt to close a compiler session.

Commit:
[ci] Tidy up workflow definitions.

Commit:
[runtime] Set Python runtime logging verbosity from debug level.

Commit:
[runtime] Fix Python shutdown signal handler.

Commit:
[tests] Treat warnings as errors in `make install-test`.

Commit:
Bump ProGraML dependency.

Commit:
[ci] Keep the $CC and $CXX variables set for install-test.

ChrisCummins committed:
Rollback grpc dependency to 1.36.0.

ChrisCummins committed:
Add missing whitespace in command.

ChrisCummins committed:
[packaging] Update curl package version for docker.

ChrisCummins committed:
[packaging] Add missing test dependencies to Dockerfile.

ChrisCummins committed:
Add zlib to list of install deps.

ChrisCummins committed:
[packaging] Add missing dependencies to Dockerfile.

ChrisCummins committed:
[service] Register SIGTERM handlers immediately.

ChrisCummins committed:
[packaging] Add a container_init script.

Commit:
[rpc] Workaround service link-time error.

ChrisCummins commented:
I have been blocked on this issue for a couple of days now and cannot seem to figure it out. I have been unable to reproduce this locally on any build environment, including the docker environment used for building the linux wheels. The problem affects all targets that use the new C++ CompilerGymService implementation, and produces the same link-time error:

```
ERROR: /home/runner/work/CompilerGym/CompilerGym/compiler_gym/envs/llvm/service/BUILD:53:10: Linking compiler_gym/envs/llvm/service/compiler_gym-llvm-service-prelinked failed: (Exit 1): clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/1570/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/compiler_gym-llvm-service-prelinked-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/1570/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/compiler_gym-llvm-service-prelinked-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

```
ERROR: /home/runner/work/CompilerGym/CompilerGym/examples/example_compiler_gym_service/service_cc/BUILD:7:10: Linking examples/example_compiler_gym_service/service_cc/compiler_gym-example-service-cc failed: (Exit 1): clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2646/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/compiler_gym-example-service-cc-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2646/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/compiler_gym-example-service-cc-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

```
ERROR: /home/runner/work/CompilerGym/CompilerGym/examples/BUILD:120:10: Linking examples/RandomSearch failed: (Exit 1): clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2798/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/RandomSearch-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2798/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/RandomSearch-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

ChrisCummins committed:
[tests] Relax strictness of error message checks.

ChrisCummins committed:
[Makefile] Fix install-test working directory path for coverage.

ChrisCummins committed:
[ci] Split test jobs into multiple workflows.

Try and keep all jobs in a workflow the same. E.g. don't bundle code
coverage along with unit tests, etc.

ChrisCummins commented:
Fixed the destructor linker error by rewriting the code using for loops rather than direct vector assignment. There's still a few bits to fix up in the CI configs then I'll merge.

ChrisCummins committed:
[service] Disable coverage in subprocess script.

ChrisCummins committed:
[tests] Add a verbose logging test for example environments.

ChrisCummins committed:
[tests] Move random actions test into fuzz directory.

ChrisCummins commented:
All test signals green. Merging!
